op-proposer/proposer/l2_output_submitter.go

```diff
@@ -321,6 +322,40 @@ func proposeL2OutputTxData(abi *abi.ABI, output *eth.OutputResponse) ([]byte, er
 
 // sendTransaction creates & sends transactions through the underlying transaction manager.
 func (l *L2OutputSubmitter) sendTransaction(ctx context.Context, output *eth.OutputResponse) error {
+	// We wait until l1head is greater than l1blocknum before sending the proposal tx, otherwise it
+	// will immediately fail when checking the l1 blockhash. This is because EstimateGas uses
+	// "latest" state to execute the transaction by default, meaning inside the call, the head
+	// block is considered "pending" instead of committed. In the case l1blocknum == l1head then,
+	// blockhash(l1blocknum) will produce a value of 0 within EstimateGas, and the call will fail
+	// when the contract checks that l1blockhash matches blockhash(l1blocknum).
+	//
+	// We actually require l1head > (l1blocknumhead+1), otherwise we still see sporadic Send
+	// errors, presumably because the head of the chain is not sufficiently stable.
+	//
+	// TODO: Consider forcing EstimateGas to use pending state by specifying block number == -1 in
+	// the RPC call. This unfortunately is not currently supported by ethclient.
```

Comment:
I think I actually prefer your change to using `pending`. It avoids the potential for other tx in the mempool interfering with the gas estimate and improves client compatibility (not all clients support `pending`). So I'd just remove this TODO.

Reply:
Agree!

Waiting logic could be extracted into a method like `waitForL1Head(ctx, output.Status.HeadL1.Number+1)`.

Reply:
Removed the TODO and moved logic into waitForL1Head